testcases.py

```diff
@@ -692,6 +692,87 @@ def check(self) -> TestResult:
         return TestResult.SUCCEEDED
 
 
+class TestCaseAmplificationLimit(TestCase):
+    @staticmethod
+    def name():
+        return "amplificationlimit"
+
+    @staticmethod
+    def testname(p: Perspective):
+        return "transfer"
+
+    @staticmethod
+    def abbreviation():
+        return "A"
+
+    @staticmethod
+    def desc():
+        return "The server obeys the 3x amplification limit."
+
+    def certs_dir(self):
+        if not self._cert_dir:
+            self._cert_dir = tempfile.TemporaryDirectory(dir="/tmp", prefix="certs_")
+            generate_cert_chain(self._cert_dir.name, 9)
+        return self._cert_dir.name + "/"
+
+    @staticmethod
+    def scenario() -> str:
+        """ Scenario for the ns3 simulator """
+        # Let the ClientHello pass, but drop a bunch of retransmissions afterwards.
+        return "droplist --delay=15ms --bandwidth=10Mbps --queue=25 --drops_to_server=2,3,4,5,6,7"
+
+    def get_paths(self):
+        self._files = [self._generate_random_file(5 * KB)]
+        return self._files
+
+    def check(self) -> TestResult:
+        if not self._keylog_file():
+            logging.info("Can't check test result. SSLKEYLOG required.")
+            return TestResult.UNSUPPORTED
+        num_handshakes = self._count_handshakes()
+        if num_handshakes != 1:
+            logging.info("Expected exactly 1 handshake. Got: %d", num_handshakes)
+            return TestResult.FAILED
+        if not self._check_version_and_files():
+            return TestResult.FAILED
+        # Check the highest offset of CRYPTO frames sent by the server.
+        # This way we can make sure that it actually used the provided cert chain.
+        max_handshake_offset = 0
+        for p in self._server_trace().get_handshake(Direction.FROM_SERVER):
+            if hasattr(p, "crypto_offset"):
+                max_handshake_offset = max(
+                    max_handshake_offset, int(p.crypto_offset) + int(p.crypto_length)
+                )
+        if max_handshake_offset < 7500:
+            logging.info(
+                "Server sent too little Handshake CRYPTO data (%d bytes). Not using the provided cert chain?",
+                max_handshake_offset,
+            )
+            return TestResult.FAILED
+        logging.debug(
+            "Server sent %d bytes in Handshake CRYPTO frames", max_handshake_offset
+        )
+        # Check that the server didn't send more than 3x what the client sent.
+        client_packets = self._server_trace().get_raw_packets(Direction.FROM_CLIENT)
+        second_packet_send_time = client_packets[1].sniff_time
+        first_packet_size = int(client_packets[0].ip.len)
+        server_sent = 0
+        for p in self._server_trace().get_raw_packets(Direction.FROM_SERVER):
+            if p.sniff_time > second_packet_send_time:
+                break
+            server_sent += int(p.ip.len)
```

Comment:
I'm not sure this is quite right. What if the client sends two packets back-to-back? I think what you want to do is go through the trace, including packets in both direction, and do something like the following (in pseudo code):
```
allowed = 0
for line in trace:
    if packet.FROM_CLIENT:
        if packet is HANDSHAKE:
            return TestResult.SUCCESS
        allowed += 3 * packet.size()
    if packet.FROM_SERVER:
        if packet.size() > allowed:
            return TestResult.FAILED
        else
            allowed -= packet.size()
```

Reply:
We use `--drops_to_server=2,3,4,5,6,7`, so even if the client sends a bunch of ClientHellos, the server will only receive a single packet for the first 2 seconds of the connection or so.
The logic here makes sure that the server doesn't go above the 3x-limit during that time.

I didn't implement checks for packets sent later, as
1. that would require more book-keeping and
2. the single packet received by the server would be the primary use case for a reflection attack (as it's fire-and-forget for the attacker)

Reply:
Actually, re-reading your pseudo-code, this totally makes sense to me now. It's more precise, and it doesn't require more book-keeping. Will update the PR.